### PR TITLE
EmitMessages Effect implementation

### PR DESCRIPTION
**Pull request by marcin-cebo**

### CODE REVIEW

File: src/main/kotlin/com/pubnub/api/subscribe/eventengine/effect/EmitMessagesProvider.kt [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+import com.pubnub.api.models.consumer.PNStatus
+import com.pubnub.api.models.consumer.pubsub.PNMessageResult
+import com.pubnub.api.models.consumer.pubsub.PNPresenceEventResult
+import com.pubnub.api.models.consumer.pubsub.PNSignalResult
+import com.pubnub.api.models.consumer.pubsub.files.PNFileEventResult
+import com.pubnub.api.models.consumer.pubsub.message_actions.PNMessageActionResult
+import com.pubnub.api.models.consumer.pubsub.objects.PNObjectEventResult
+
+interface EmitMessagesProvider {
```
Comment (kleewho) [inline src/main/kotlin/com/pubnub/api/subscribe/eventengine/effect/EmitMessagesProvider.kt:11]: The name `Provider` is usually associated with something that returns something. None of the methods in it return anything so I don't think it should be named `Provider` at all

File: src/main/kotlin/com/pubnub/api/subscribe/eventengine/effect/EmitMessagesProvider.kt [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+    fun announce(presence: PNPresenceEventResult) // toDo move to Presence EventEngine
+
+    fun announce(signal: PNSignalResult)
+
+    fun announce(messageAction: PNMessageActionResult)
+
+    fun announce(pnObjectEventResult: PNObjectEventResult)
+
+    fun announce(pnFileEventResult: PNFileEventResult)
```
Comment (kleewho) [inline src/main/kotlin/com/pubnub/api/subscribe/eventengine/effect/EmitMessagesProvider.kt:24]: Another question one might have is do we actually want to have this distinction on method level? Do we need so many methods named the same or maybe just one would be enough? 

```
    fun betterMethodNameThenAnnounceMaybe(event: PNEvent)
```

File: src/test/kotlin/com/pubnub/api/subscribe/eventengine/effect/BaseEffectTest.kt [hunk 3]
```diff
@@ -5,8 +5,8 @@ import com.pubnub.api.models.consumer.pubsub.BasePubSubResult
 import com.pubnub.api.models.consumer.pubsub.PNEvent
 import com.pubnub.api.models.consumer.pubsub.PNMessageResult
 
-open class BaseEffectTest {
-    protected fun createMessages(): List<PNEvent> {
+abstract class BaseEffectTest {
```
Comment (kleewho) [inline src/test/kotlin/com/pubnub/api/subscribe/eventengine/effect/BaseEffectTest.kt:8]: Do we need a class at all? Why not just make the `createPNMessageResultList` just a function?

File: src/test/kotlin/com/pubnub/api/subscribe/eventengine/effect/EmitMessagesEffectTest.kt [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+        pnObjectEventResult: PNObjectEventResult
+    ): List<PNEvent> {
+        val listOfEvents = mutableListOf<PNEvent>()
+        listOfEvents.add(pnMessageResult)
+        listOfEvents.add(pnFileEventResult)
+        listOfEvents.add(pnSignalResult)
+        listOfEvents.add(pnMessageActionResult)
+        listOfEvents.add(pnObjectEventResult)
+
+        return listOfEvents
```
Comment (kleewho) [inline src/test/kotlin/com/pubnub/api/subscribe/eventengine/effect/EmitMessagesEffectTest.kt:156]:
```suggestion
        return listOf(
            pnMessageResult,
            pnFileEventResult,
            pnSignalResult,
            pnMessageActionResult,
            pnObjectEventResult
        )
```

**Review by kleewho — CHANGES_REQUESTED** [review]

File: src/main/kotlin/com/pubnub/api/eventengine/EffectDispatcher.kt [hunk 5]
```diff
@@ -24,14 +19,14 @@ class EffectDispatcher<T : EffectInvocation>(
 
             is Managed -> {
                 managedEffects.remove(effectInvocation.id)?.cancel()
-                val managedEffect = managedEffectFactory.create(effectInvocation) ?: return
-                managedEffects[effectInvocation.id] = managedEffect
+                val managedEffect = effectFactory.create(effectInvocation) ?: return
+                managedEffects[effectInvocation.id] = managedEffect as ManagedEffect
```
[the remark this answers is not in the document]
Comment (marcin-cebo) [reply src/main/kotlin/com/pubnub/api/eventengine/EffectDispatcher.kt:23]: My understanding is that each invocation has type and code inside  "Managed -> { } " is executed if Invocation has type "Managed".  SubscribeManagedEffectFactory creates Effects and knows how to map invocation to effect of proper type.

To double check we could enhance unit test    

    @Test
    fun `should return handshake effect when getting Handshake invocation`() {
        // when
        val managedEffect =
            subscribeManagedEffectFactory.create(SubscribeEffectInvocation.Handshake(channels, channelGroups))

        // then
        assertThat(managedEffect, instanceOf(HandshakeEffect::class.java))
        assertThat(managedEffect, instanceOf(ManagedEffect::class.java)). <-- add this line to check  type of effect
    }

File: src/main/kotlin/com/pubnub/api/subscribe/eventengine/effect/EmitMessagesProvider.kt (hunk 1 above)
Comment (marcin-cebo) [reply src/main/kotlin/com/pubnub/api/subscribe/eventengine/effect/EmitMessagesProvider.kt:11]: ok, so maybe EmitMessagesConsumer ?

File: src/test/kotlin/com/pubnub/api/subscribe/eventengine/effect/BaseEffectTest.kt (hunk 3 above)
Comment (marcin-cebo) [reply src/test/kotlin/com/pubnub/api/subscribe/eventengine/effect/BaseEffectTest.kt]: I wanted do have function that will be used by many classes.
Is static method better design that having abstract class with method like here?

File: src/test/kotlin/com/pubnub/api/subscribe/eventengine/effect/EmitMessagesEffectTest.kt (hunk 4 above)
Comment (marcin-cebo) [reply src/test/kotlin/com/pubnub/api/subscribe/eventengine/effect/EmitMessagesEffectTest.kt]: I like this. I will make a change.

File: src/main/kotlin/com/pubnub/api/subscribe/eventengine/effect/EmitMessagesProvider.kt (hunk 2 above)
Comment (marcin-cebo) [reply src/main/kotlin/com/pubnub/api/subscribe/eventengine/effect/EmitMessagesProvider.kt]: I though that when putting embedding EE into SDK we could have ListenerManager implemented this interface. It has all those method implemented but doesn't have method announce(event: PNEvent)
What do you think?

**Pull request closed without merging**